### ISSUE TITLE
Locale-based 'printf %T' output is broken

### ISSUE THREAD

**Issue by McDutchie** [ksh93/ksh#52]
E.g. on macOS, as well as FreeBSD:

```
$ LANG=nl_NL.UTF-8 TZ=Europe/Helsinki arch/*/bin/ksh -c 'date; printf %T\\n now'
di 30 jun 2020 03:49:12 EEST
di 30 jun 03:49:12 2020
```

(the time zone is missing from `printf %T` output and the year is printed in the wrong order)

And on Linux (Debian):

```
$ LANG=nl_NL.UTF-8 TZ=Europe/Helsinki arch/*/bin/ksh -c 'date; printf %T\\n now'
di 30 jun 2020  3:50:40 EEST
di 30 jun 2020 di jun 30  EEST 2020:50:40 EEST
```

That different systems produce differently corrupted output is interesting.

Related: #6, #38, #62

**Comment by McDutchie** [comment]
> E.g. on macOS, as well as FreeBSD:
> 
> ```
> $ LANG=nl_NL.UTF-8 TZ=Europe/Helsinki arch/*/bin/ksh -c 'date; printf %T\\n now'
> di 30 jun 2020 03:49:12 EEST
> di 30 jun 03:49:12 2020
> ```
> 
> (the time zone is missing from `printf %T` output and the year is printed in the wrong order)

The macOS behaviour is not a bug in ksh, possibly a bug in macOS's locale data. The AST tmxfmt() function calls tmlocale() to get a date format string from the operating system. So this can vary by OS. If we apply this diff:
```diff
diff --git a/src/lib/libast/tm/tmxfmt.c b/src/lib/libast/tm/tmxfmt.c
index 5023d43..1e93df2 100644
--- a/src/lib/libast/tm/tmxfmt.c
+++ b/src/lib/libast/tm/tmxfmt.c
@@ -128,6 +128,7 @@ tmxfmt(char* buf, size_t len, const char* format, Time_t t)
 	if (!format || !*format)
 		format = tm_info.deformat;
 	oformat = format;
+sfprintf(sfstderr,"[DEBUG] format = [%s]\n",format);
 	flags = tm_info.flags;
 	sp = &stack[0];
 	cp = buf;
```
we see that the 'printf %T' output corresponds to the format string:
```
$ LANG=C TZ=Europe/Helsinki arch/*/bin/ksh -c 'date; printf %T\\n now'
Fri Jan 22 08:02:00 EET 2021
Fri Jan 22 08:02:00 EET 2021
[DEBUG] format = [%a %b %e %T %Z %Y]
$ LANG=nl_NL.UTF-8 TZ=Europe/Helsinki arch/*/bin/ksh -c 'date; printf %T\\n now'
vr 22 jan 2021 08:02:02 EET
vr 22 jan 08:02:02 2021
[DEBUG] format = [%a %e %b %X %Y]
```

**Comment by McDutchie** [comment]
> And on Linux (Debian):
> 
> ```
> $ LANG=nl_NL.UTF-8 TZ=Europe/Helsinki arch/*/bin/ksh -c 'date; printf %T\\n now'
> di 30 jun 2020  3:50:40 EEST
> di 30 jun 2020 di jun 30  EEST 2020:50:40 EEST
> ```

This was legitimately a bug -- corrupted output. However, I can no longer reproduce it on current 93u+m running on Debian (still reproducible on 93u+).

**Comment by McDutchie** [comment]
The bug on Debian (and other Linux) was fixed long ago by this commit: 9526b3fa08

Which makes sense. It was not data corruption at all, it was an incompatibility in the format string between GNU/Linux locale data and AST. The format string for the Dutch locale on Linux is "`%a %e %b %Y %k:%M:%S %Z`", and `%k` is precisely one of the nonstandard AST format specifiers that was changed in that commit to be compatible with Linux date(1) -- and the latter is of course what Linux locale data is written for.
